windows/RNFS/RNFSManager.cpp

```diff
@@ -0,0 +1,1115 @@
+        itemInfo["size"] = properties.Size();
+        itemInfo["type"] = item.IsOfType(StorageItemTypes::Folder) ? 1 : 0;
+
+        resultsArray.push_back(std::move(itemInfo));
+    }
+
+    promise.Resolve(resultsArray);
+}
+catch (const hresult_error& ex)
+{
+    // "Failed to read directory."
+    promise.Reject(winrt::to_string(ex.message()).c_str());
+}
+
+
+winrt::fire_and_forget RNFSManager::read(std::string filepath, uint32_t length, uint64_t position, RN::ReactPromise<std::string> promise) noexcept
+try
+{
+    winrt::hstring directoryPath, fileName;
+    splitPath(filepath, directoryPath, fileName);
+
+    StorageFolder folder{ co_await StorageFolder::GetFolderFromPathAsync(directoryPath) };
+    StorageFile file{ co_await folder.GetFileAsync(fileName) };
+
+    Streams::Buffer buffer{ length };
+
+    Streams::IRandomAccessStream stream{ co_await file.OpenReadAsync() };
+    stream.Seek(position);
+
+    stream.ReadAsync(buffer, length, Streams::InputStreamOptions::None);
+    std::string result{ winrt::to_string(Cryptography::CryptographicBuffer::EncodeToBase64String(buffer)) };
+
+    promise.Resolve(result);
+}
+catch (const hresult_error& ex)
+{
+    hresult result{ ex.code() };
+    if (result == 0x80070002) // FileNotFoundException
+    {
+        promise.Reject(RN::ReactError{ "ENOENT", "ENOENT: no such file or directory, open " + filepath });
+    }
+    else if (result == 0x80070005) // UnauthorizedAccessException
+    {
+        promise.Reject(RN::ReactError{"EISDIR", "EISDIR: Could not open file for reading" });
+    }
+    else 
+    {
+        // "Failed to read from file."
+        promise.Reject(winrt::to_string(ex.message()).c_str());
+    }
+}
+
+
+winrt::fire_and_forget RNFSManager::hash(std::string filepath, std::string algorithm, RN::ReactPromise<std::string> promise) noexcept
+try
+{
+    // Note: SHA224 is not part of winrt 
+    if (algorithm.compare("sha224") == 0)
+    {
+        promise.Reject(RN::ReactError{ "Error", "WinRT does not offer sha224 encryption." });
+        co_return;
+    }
+
+    winrt::hstring directoryPath, fileName;
+    splitPath(filepath, directoryPath, fileName);
+
+    StorageFolder folder{ co_await StorageFolder::GetFolderFromPathAsync(directoryPath) };
+    StorageFile file{ co_await folder.GetFileAsync(fileName) };
+
+    auto search{ availableHashes.find(algorithm) };
+    if (search == availableHashes.end())
+    {
+        promise.Reject(RN::ReactError{ "Error", "Invalid hash algorithm " + algorithm});
+        co_return;
+    }
+
+    CryptographyCore::HashAlgorithmProvider provider{ search->second() };
+    Streams::IBuffer buffer{ co_await FileIO::ReadBufferAsync(file) };
+
+    auto hashedBuffer{ provider.HashData(buffer) };
+    auto result{ winrt::to_string(Cryptography::CryptographicBuffer::EncodeToHexString(hashedBuffer)) };
+
+    promise.Resolve(result);
+}
+catch (const hresult_error& ex)
+{
+    hresult result{ ex.code() };
```

Comment:
extract this out to its own helper function and calling it from the different catch clauses so you don't repeat the same code (and ensures the string messages are consistent as well as the set of errors handled)